pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -0,0 +1,402 @@
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, bool, error) {
+	return spec, false, nil
+}
+
+func machineSpecParsingTerminalError(err error) error {
+	return cloudprovidererrors.TerminalError{
+		Reason:  v1alpha1.InvalidConfigurationMachineError,
+		Message: fmt.Sprintf("Failed to parse MachineSpec due to %v", err),
+	}
+}
+
+func getClient(username, password, address string, allowInsecure bool) (*govmomi.Client, error) {
+	clientUrl, err := url.Parse(fmt.Sprintf("%s/sdk", address))
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse vsphere url due to %v", err),
+		}
+	}
+	clientUrl.User = url.UserPassword(username, password)
+
+	client, err := govmomi.NewClient(context.TODO(), clientUrl, allowInsecure)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to get vsphere client due to %v", err),
+		}
+	}
+	return client, nil
+}
+
+func (p *provider) getConfig(s runtime.RawExtension) (*Config, *providerconfig.Config, error) {
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	rawConfig := RawConfig{}
+	err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig)
+
+	c := Config{}
+	c.TemplateVMName, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.TemplateVMName)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	c.Username, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Username)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	c.Password, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Password)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	c.VSphereURL, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.VSphereURL)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	c.Datacenter, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Datacenter)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	c.Cluster, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Cluster)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	c.Datastore, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Datastore)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	c.AllowInsecure, err = p.configVarResolver.GetConfigVarBoolValue(rawConfig.AllowInsecure)
+	if err != nil {
+		return nil, nil, machineSpecParsingTerminalError(err)
+	}
+
+	c.CPUs = rawConfig.CPUs
+	c.MemoryMB = rawConfig.MemoryMB
+
+	return &c, &pconfig, nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
```

Comment:
Ah okay, this is pretty confusing tbh because it means there are some places in a provider where we must return TerminalErrors and some where we must not. Also it sounds to me as if `Create` should not return a TerminalError, because `Validate` contains all possible validations already and if create fails, the controller will retry which means it will first execute `Validate`.

What do you think about keeping it as simple as possible and use `Validate` as the only func that returns TerminalErrors? We could even reflect that in its signature.

Reply:
I really like this idea. I think we could follow this path only if we are sure that things are constant and don't change in the system. That is, there are not updates to manifest, user credentials and things inside cloud provider after a machine has been created (we called `Validate` method) and before we call `Delete` method. I am not sure if we can guarantee that ?

Reply:
Well, we could run `Validate` under some more conditions, e.G.:

* `Get` fails and the error is not `instanceNotFound`
* `Delete` fails (Not sure if that makes sense)

But if we do that it should IMHO be a separate PR.

For this PR, is it okay if I:

* Remove the terminalErrors from Validate
* Thror them in `Create` only if appropriate

?

Reply:
> For this PR, is it okay if I:
> Remove the terminalErrors from Validate
> Throw them in only Create if appropriate

yes.